### ISSUE TITLE
NullInjectorError: No provider for InjectorRef!

### ISSUE THREAD

**Issue by skorunka** [robisim74/angular-l10n#146]
**I'm submitting a...**
```
[x] support request
```

```
Unhandled Promise rejection: StaticInjectorError[InjectorRef]: 
  StaticInjectorError[InjectorRef]: 
    NullInjectorError: No provider for InjectorRef! ; Zone: angular ; Task: Promise.then ; Value: Error: StaticInjectorError[InjectorRef]: 
  StaticInjectorError[InjectorRef]: 
    NullInjectorError: No provider for InjectorRef!
    at _NullInjector.get (core.js:923)
    at resolveToken (core.js:1211)
    at tryResolveToken (core.js:1153)
    at StaticInjector.get (core.js:1024)
    at resolveToken (core.js:1211)
    at tryResolveToken (core.js:1153)
    at StaticInjector.get (core.js:1024)
    at resolveNgModuleDep (core.js:10585)
    at NgModuleRef_.get (core.js:11806)
    at resolveNgModuleDep (core.js:10585) Error: StaticInjectorError[InjectorRef]: 
  StaticInjectorError[InjectorRef]: 
    NullInjectorError: No provider for InjectorRef!
    at _NullInjector.get (webpack-internal:///../../../core/esm5/core.js:1119:19)
    at resolveToken (webpack-internal:///../../../core/esm5/core.js:1407:24)
    at tryResolveToken (webpack-internal:///../../../core/esm5/core.js:1349:16)
    at StaticInjector.get (webpack-internal:///../../../core/esm5/core.js:1220:20)
    at resolveToken (webpack-internal:///../../../core/esm5/core.js:1407:24)
    at tryResolveToken (webpack-internal:///../../../core/esm5/core.js:1349:16)
    at StaticInjector.get (webpack-internal:///../../../core/esm5/core.js:1220:20)
    at resolveNgModuleDep (webpack-internal:///../../../core/esm5/core.js:10781:25)
    at NgModuleRef_.get (webpack-internal:///../../../core/esm5/core.js:12002:16)
    at resolveNgModuleDep (webpack-internal:///../../../core/esm5/core.js:10781:25)
api.onUnhandledError @ zone.js:690
handleUnhandledRejection @ zone.js:717
_loop_1 @ zone.js:707
api.microtaskDrainDone @ zone.js:711
drainMicroTaskQueue @ zone.js:610
ZoneTask.invokeTask @ zone.js:503
invokeTask @ zone.js:1540
globalZoneAwareCallback @ zone.js:1566
```

I got this error while loading one of my modules. I just spotted, the `InjectorRef` is part of `angular-l10n` project. Any idea what should I focus on? Thank you.

**Comment by robisim74** [comment]
Are you getting this error only in one module? Do other modules work?

_InjectorRef_ is provided by _TranslationModule_ or _LocalizationModule_ through _forRoot_ and _forChild_ methods. Is your module a sub module of _AppModule_?

**Comment by skorunka** [comment]
Hi, I do get this error in a lazy loaded sub-module. Other, lazy-loaded sub-module, works.

**Comment by robisim74** [comment]
You can get this error if the module or parent modules are not invoking the _forRoot_ or _forChild_ method of _TranslationModule_ or _LocalizationModule_.

Try to do the following check.

Add manually to your module the service:

```
providers: [
        ...
        InjectorRef
    ],
```

Then, you should get something like this:

```
Error: StaticInjectorError[L10nLoader]: 
  StaticInjectorError[L10nLoader]: 
    NullInjectorError: No provider for L10nLoader!
```

**Comment by skorunka** [comment]
Hmm, it did not help. That's weird. Still the same error.

I have one `SharedModule`, where I do `imports:[LocalizationModule]` & `exports[LocalizationModule]` then in `AppModule` I do `imports:[LocalizationModule.forRoot(defaultL10nConfig)]`.  In lazy loaded modules I do:
```
@NgModule({
	imports: [
		SharedModule.forRoot(),
```

A simple module works (no above exceptiona) but a bit more complex one does, not. I think I have to remove dependencies in the module which doe snot work until it works and then add one by one and see where the problem is.

**Comment by robisim74** [comment]
What is a _simple module_ and a _complex_ module? Could you post the essential of your shared module and your routing module? If you can, please provide a minimal repro.

**Comment by skorunka** [comment]
It is not easy to create a repro, otherwise I would do that. I think there also might be something with migrating to a new version. I will try to remove any ussage(pipes, services) of this library and see if it helps and then add one by one.

**Comment by robisim74** [comment]
Ok, however is weird that you have a forRoute method in a shared module (it is usually used for root modules, for example into libraries).

See also the following recent Angular issues for this kind of error: https://github.com/angular/angular/search?q=StaticInjectorError&type=Issues&utf8=%E2%9C%93

**Comment by skorunka** [comment]
I don't have `LocalizationModule.forRoot(defaultL10nConfig)` in `SharedModule` if it is what you mean. ;)

**Comment by Collaborator12** [comment]
Dear @robisim74 ,

First of all, I would like to say thank you ! for such a great library  :+1: 

I am also getting same error as @skorunka .

My scenario is bit different, as follows

1) I have created a component library using ng-packagr & am using your translation module into it. It is a angular cli app. When I run the library on it's own, it perfectly shows translation through TranslateModule.
2) But  I want to use components from this library into any other angular cli project & translate them as per requirement of that new cli project. This is where I am getting error as mentioned by @skorunka .
3) I have components that are divided into feature modules & configuration for TranslateModule is only in AppModule. Through a Shared module I use it in all my feature modules.
4) I import the feature module from library into new  cli project when I want to use component under that feature module. Should I be using forRoot somewhere in order to ensure that I can provide translations from new cli project for the imported components ?

Please help me @robisim74 !!

**Comment by robisim74** [comment]
@Abhishek-Kanitkar 

It's not possible for me recreate the problem without a repro.

As I said, `InjectorRef` is the first class injected by this library. The classes of this library are injected only when you call the `forRoot` or `forChild` methods: https://github.com/robisim74/angular-l10n/blob/master/src/modules/translation.module.ts#L35 

The main reason why you get this error is that _the module or the parent module_ in your host app is never calling those methods, so the services are not instantiated. Indeed, the Angular modules are a tree, and the instance of an injectable class is visible only in the child modules.

Another reason could be if you are using `npm link` (it is a known issue), usually solved by adding `preserve symlinks` to your Angular CLI project (you can find a lot of these situation in official Angular's repositories).

**Comment by Collaborator12** [comment]
@robisim74 

Lets assume that in my host app I don't want to provide any translations through TranslateModule, then whatever translations provided in library should work right ? I don't 'necessarily' have to add TranslateModule.forRoot() into every host app that tries to use my component library.

So in above case can the error be avoided ?

Following is my shared.module.ts

```
import { NgModule, ModuleWithProviders } from '@angular/core';
import { TranslationModule } from 'angular-l10n';

const sharedModules: any[] = [
    TranslationModule
];

@NgModule({
    imports: sharedModules,
    exports: sharedModules
})
export class SharedModule {
	
}
```
Following is my footer.module.ts

```
import {ModuleWithProviders, NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {FooterComponent} from './footer.component';
import {HttpClient, HttpClientModule} from '@angular/common/http';
import {BrowserModule} from '@angular/platform-browser';
import { SharedModule } from "../shared/shared.module";

@NgModule({
  imports: [
    SharedModule,
    CommonModule,
    BrowserModule,
    HttpClientModule    
  ],
  declarations: [FooterComponent],
  providers: [],
  exports: [FooterComponent],
})
export class FooterModule {
  
}

```
footer.component.html

```
<p>{{ 'Greeting' | translate:lang }}</p>

        <p title="{{ 'Greeting' | translate:lang }}">{{ 'Subtitle' | translate:lang }}</p>
```
app.module.ts

```
import {BrowserModule} from '@angular/platform-browser';
import {NgModule} from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import {AppComponent} from './app.component';
import { L10nConfig, L10nLoader, TranslationModule, StorageStrategy, ProviderType } from 'angular-l10n';
import { SharedModule } from "./modules/shared/shared.module";
import { FooterModule } from "./modules/footer/footer.module";

const l10nConfig: L10nConfig = {
    locale: {
        languages: [
            { code: 'en', dir: 'ltr' },
            { code: 'it', dir: 'ltr' }
        ],
        language: 'en',
        storage: StorageStrategy.Cookie
    },
    translation: {
        providers: [
            { type: ProviderType.Static, prefix: './assets/locale-' }
        ],
        caching: true,
        missingValue: 'No key'
    }
};

@NgModule({
  declarations: [
    AppComponent,
  ],
  imports: [
    FooterModule,
  	SharedModule.forRoot(),
    HttpClientModule,
    BrowserModule,
    TranslationModule.forRoot(l10nConfig)
  ],
  providers: [],
  bootstrap: [AppComponent],
})

export class AppModule {
	constructor(public l10nLoader: L10nLoader) {
        this.l10nLoader.load();
    }
}

```

===============================================================

In my host app

app.module.ts

```
import {NgModule} from '@angular/core';

import {AppComponent} from './app.component';
import {HttpClientModule, HttpClient} from '@angular/common/http';
import {CommonModule} from '@angular/common';
import {BrowserModule} from '@angular/platform-browser';
import {FooterModule} from 'lib-playground';

@NgModule({
  declarations: [
    AppComponent,
  ],
  imports: [
    CommonModule,
    BrowserModule,
    HttpClientModule,
    FooterModule    
  ],
  providers: [],
  bootstrap: [AppComponent],
})

export class AppModule {
}

```

app.component.html

`<bd-footer></bd-footer>`


Like I said above, every host app that is using my component library, may not provide translations by configuring TranslateModule.

In that case default translations should work right ?

So what am I missing here ?

In my host app, I am just importing the FooterModule to import FooterComponent.

I have provided translations in library, but not necessary that I will provide translations in host app as well.

Can you please help point out what I am missing here ?


Basically I have 2 Angular Cli apps.

1) One app is converted into a library using ng-packagr. It has translations using TranslateModule

2) Second app imports the first app through npm install (library name.tgz file). It may not provide any translations.

Please help !

**Comment by robisim74** [comment]
@Abhishek-Kanitkar I just tried to create a library (using a my starter), and to import it into an Angular CLI app, and it works without problems: just import the library and the component into the app.

So I don't know what happens in your project: I never used ng-packagr to convert an app into a library, but however where do your host app starts your `lib-playground`? In the AppModule you posted I don't see that your library is imported (only the FooterModule is not sufficient). If your library does not start, neither this library starts.

**Comment by robisim74** [comment]
I'm closing this issue due to lack of repro. Feel free to reopen if you need. Greetings

**Comment by macwier** [comment]
For anyone interested in this.
Instead of doing normal `npm link`, I've solved this kind of problem with editing tsconfig.json (in the application workspace) and adding:
- The library with absolute path
- angular/* pointing to node_modules

I.e.
```json
        "paths": {
            "my-library": [
                "C:\\repos\\my-library-app\\dist\\my-library"
            ],
            "@angular/*": [
                "node_modules/@angular/*"
            ],
```

Might be that all peer dependencies should be added like the `@angular`, but in the end this was enought to get rid of the error.
